File: base_demo/src/topic_sub.cpp
#include <chrono>
#include <memory>
#include "rclcpp/rclcpp.hpp"
#include "base_demo/msg/robot_info.hpp"

using std::placeholders::_1;

class TopicSubscriber : public rclcpp::Node
{
public:
  TopicSubscriber()
  : Node("topic_sub")
  {
    RCLCPP_INFO(this->get_logger(), "topic_sub node is Ready!");
    subscription_ = this->create_subscription<base_demo::msg::RobotInfo>(
      "robot_info", 10, std::bind(&TopicSubscriber::topic_callback, this, _1));
  }

private:
  void topic_callback(const base_demo::msg::RobotInfo & msg) const 
  {
    if(msg.pose.position.x == 20.0){
    RCLCPP_INFO(this->get_logger(), "The robot has reached the target pose.");
      if(msg.is_carry){
        RCLCPP_INFO(this->get_logger(), "The robot is carrying objects");
      }else{
      RCLCPP_INFO(this->get_logger(), "The robot is not carrying objects");
      }
    }else{
      RCLCPP_INFO(this->get_logger(), msg.state.c_str());
      RCLCPP_INFO(this->get_logger(), "Robot pose x : %.2fm; y : %.2fm; z : %.2fm", msg.pose.position.x, msg.pose.position.y, msg.pose.position.z);
    }
  }
  rclcpp::Subscription<base_demo::msg::RobotInfo>::SharedPtr subscription_; 

};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<TopicSubscriber>());
  rclcpp::shutdown();
  return 0;
}

File: base_demo/scripts/topic_sub.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
import rclpy
from rclpy.node import Node
from base_demo.msg import RobotInfo


class TopicSubscriber(Node):

    def __init__(self):
        super().__init__('topic_subscriber')
        self.get_logger().info('topic_sub python node is Ready!')

        self.subscription = self.create_subscription(
            RobotInfo,
            'robot_info',
            self.listener_callback,
            10)
        self.subscription  # prevent unused variable warning

    def listener_callback(self, data):
        if data.pose.position.x == 20:
            self.get_logger().info('The robot has reached the target pose.')
            if data.is_carry:
                self.get_logger().info('The robot is carrying objects')
            else:
                self.get_logger().info('The robot is not carrying objects')
        else:
            self.get_logger().info(data.state)
            self.get_logger().info('Robot pose x : %.2fm' %data.pose.position.x )



def main(args=None):
    rclpy.init(args=args)

    topic_subscriber = TopicSubscriber()

    rclpy.spin(topic_subscriber)

    # Destroy the node explicitly
    # (optional - otherwise it will be done automatically
    # when the garbage collector destroys the node object)
    topic_subscriber.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

